Fix: Ensure Related Products collection displays correctly in Single Product template

imanish003 opened the pull request:
### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR fixes an issue with the Related Products collection not displaying on Single Product Templates in WooCommerce 9.5 and later. The issue stems from PR #52580 which introduced the `relatedBy` attribute, but the [related-products.php](https://github.com/woocommerce/woocommerce/blob/c012af2752f6b09c53b234667995148c9316ef88/plugins/woocommerce/patterns/related-products.php) pattern was not updated to include this required attribute.

Currently, merchants need to manually fix this by either:
1. Removing and re-adding the "Related Products" collection, which sets the default values for `categories` and `tags` to true
2. Manually enabling the "Categories" and "Tags" checkboxes in the Inspector controls for an existing "Related Products" collection

The issue likely went unreported because merchants were able to work around it through the UI, but this fix ensures the feature works correctly out of the box.

**Fixes made in this PR:**

1. Updated the `related-products.php` pattern to include the `relatedBy` attribute with both `categories` and `tags` set to `true` by default.
2. Modified the `HandlerRegistry.php` file to: Set default values for `relatedBy` to `true` for both categories and tags when the attribute is not provided

These changes ensure that:
- The Related Products collection works correctly out of the box with the default pattern
- Existing blocks without the `relatedBy` attribute will show related products based on both categories and tags
- The behavior remains backwards compatible while fixing the core issue

Closes # .

(For Bug Fixes) Bug introduced in PR # .

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|        |       |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

1. Go to Single Product Template
2. Reset template to default
3. After resetting, verify that checkboxes for "Categories" and "Tags" are checked in the Inspector controls by default. (As shown in screenshot below)
![image](https://github.com/user-attachments/assets/02cfd313-5634-4dca-8475-25c16b5ae217)

4. Save and go to frontend of any product
5. Verify that Related Products are shown for currently viewed product

**Verify backwards compatibility:**

1. Using the Code Editor in the Single Product Template, manually remove the `relatedBy` attribute from the "Related Products" collection to simulate the previous default behavior
2. Save the template and view any product page on the frontend
3. Confirm that Related Products are still displayed correctly for the current product, validating the backwards compatibility

![image](https://github.com/user-attachments/assets/9e4cef88-8587-4eca-9ee6-d9698ebcdc4f)


<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Fix: Ensure Related Products collection displays correctly in Single Product template
</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

kmanijak reviewed (approved):
Works as expected:

- Related Products are displayed as expected
    - ✅ before resetting - backwards compatibility works as expected. I actually wanted to suggested the same default values before checking the code. Amazing you've already done that, great thinking!
    - ✅ after resetting - the `relatedBy` is added as expected.

imanish003 committed:
Fix: Ensure Related Products collection displays correctly in Single Product template

The Related Products collection block was not displaying products correctly in the Single Product template due to missing default values for the `relatedBy` attribute. This fix ensures that:

1. The default pattern includes the `relatedBy` attribute with both categories and tags enabled
2. Existing blocks without the `relatedBy` attribute default to showing products related by both categories and tags
3. The behavior remains backwards compatible while fixing the core functionality

This change improves the out-of-box experience for merchants by:
- Eliminating the need to manually fix Related Products collections
- Ensuring proper display without requiring UI intervention
- Maintaining expected behavior for existing installations

imanish003 commented:
Hey @kmanijak, As discussed on Slack (p1742908908806499/1742904908.890679-slack-C01DT6U03HC), I will be raising CFE for this PR. Therefore, I changed the base branch to a release branch. Would you mind testing this PR again to make sure that everything is still functioning as expected? I have also tested this PR again after the rebase, and everything seems to be working as expected. Thanks 🙏🏻

kmanijak reviewed (approved):
> Would you mind testing this PR again to make sure that everything is still functioning as expected? I have also tested this PR again after the rebase, and everything seems to be working as expected. Thanks 🙏🏻

Thanks @imanish003! I retested the changes and all looks good:
- ✅ if there's no `relatedBy` both categories and tags are taken into account by default
- ✅  when resetting the template both are enabled by default
- ✅ If I disable both of them manually ( `"relatedBy":{"categories":false,"tags":false}` ), then Related Products are not displayed (this is different than first case, becase in first case the whole property doesn't exist).

Thanks for aiming for best quality and eventually raising CFE even though I discouraged it! 💪

imanish003 committed:
Merge branch 'release/9.8' into fix/related-products-collection-not-working-on-single-product-template

imanish003 committed:
Merge branch 'release/9.8' into fix/related-products-collection-not-working-on-single-product-template
